File: backend/src/services/userService.ts
import { User } from '@prisma/client';
import { prisma } from '../config/database.js';
import { PasswordUtils } from '../utils/passwordUtils.js';

export interface CreateUserInput {
  email: string;
  name: string;
  password: string;
}

export interface LoginInput {
  email: string;
  password: string;
}

/**
 * User service handles all user-related business logic
 * Implements secure authentication with password hashing
 */
export class UserService {
  /**
   * Create a new user with hashed password
   * @param userData - User registration data
   * @returns Promise<User> - Created user (without password hash)
   * @throws Error if email already exists
   */
  async createUser(userData: CreateUserInput): Promise<Omit<User, 'passwordHash'>> {
    const existingUser = await prisma.user.findUnique({
      where: { email: userData.email },
    });

    if (existingUser) {
      throw new Error('User with this email already exists');
    }

    const passwordHash = await PasswordUtils.hashPassword(userData.password);

    const user = await prisma.user.create({
      data: {
        email: userData.email,
        name: userData.name,
        passwordHash,
      },
    });

    // Return user without password hash for security
    const { passwordHash: _, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }

  /**
   * Authenticate user login credentials
   * @param loginData - Email and password
   * @returns Promise<User | null> - User if credentials valid, null otherwise
   */
  async authenticateUser(loginData: LoginInput): Promise<Omit<User, 'passwordHash'> | null> {
    const user = await prisma.user.findUnique({
      where: { email: loginData.email },
    });

    if (!user) {
      return null;
    }

    const isPasswordValid = await PasswordUtils.verifyPassword(
      loginData.password,
      user.passwordHash
    );

    if (!isPasswordValid) {
      return null;
    }

    // Return user without password hash
    const { passwordHash: _, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }

  /**
   * Find user by ID
   * @param userId - User ID
   * @returns Promise<User | null> - User without password hash
   */
  async findById(userId: number): Promise<Omit<User, 'passwordHash'> | null> {
    const user = await prisma.user.findUnique({
      where: { id: userId },
    });

    if (!user) {
      return null;
    }

    const { passwordHash: _, ...userWithoutPassword } = user;
    return userWithoutPassword;
  }
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/service/UserService.java
package com.learningadvisor.backend.service;

import com.learningadvisor.backend.dto.AuthRequest;
import com.learningadvisor.backend.dto.UserDTO;
import com.learningadvisor.backend.entity.User;
import com.learningadvisor.backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service for user management and authentication
 * Handles user creation and password verification
 */
@Service
@Transactional
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Create a new user account
     * @param request Registration request with email, password, and name
     * @return Created user
     * @throws IllegalArgumentException if email already exists
     */
    public User createUser(AuthRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("User with email " + request.getEmail() + " already exists");
        }

        User user = User.builder()
                .email(request.getEmail())
                .name(request.getName())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .build();

        return userRepository.save(user);
    }

    /**
     * Authenticate user with email and password
     * @param request Login request with email and password
     * @return User if credentials are valid, null otherwise
     */
    public User authenticateUser(AuthRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElse(null);

        if (user != null && passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            return user;
        }

        return null;
    }

    /**
     * Find user by ID
     */
    public User findById(Long userId) {
        return userRepository.findById(userId).orElse(null);
    }

    /**
     * Convert User entity to DTO
     */
    public UserDTO toDTO(User user) {
        return UserDTO.builder()
                .id(user.getId())
                .email(user.getEmail())
                .name(user.getName())
                .createdAt(user.getCreatedAt())
                .build();
    }
}
